File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
private:
    vector<int> temp;
    vector<vector<int>> ans;
    int sum = 0;
public:
    void search(int k,int target, vector<int>& candidates){
        int n = candidates.size();
        if(sum == target){
            ans.push_back(temp);
        }else if(sum > target){
            return;
        }else{
            for(int i = k; i<n; i++){
                temp.push_back(candidates[i]);
                sum += candidates[i];
                search(i,target,candidates);
                sum -= candidates[i];
                temp.pop_back();
            }
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        search(0,target,candidates);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const res = [];
    const temp = [];
    let sum = 0;
    const search = function (k,sum) {
        if(sum > target) return;
        else if(sum === target) res.push(Array.from(temp));
        else{
            for(let i = k; i<candidates.length; i++){
                temp.push(candidates[i]);
                sum += candidates[i];
                search(i,sum);
                sum -= candidates[i];
                temp.pop();
            }
        }
    }
    search(0,sum);
    return res;
};